nep-12.mediawiki

```diff
@@ -0,0 +1,2130 @@
+    {
+      'assetID': '602c79718b16e442de58778e148d0b1084e3b2dffd5de6b7b16cee7969282de7',
+      'symbol': 'GAS',
+      'amount': '777.0001',
+    },
+    {
+      'assetID': 'c36aee199dbba6c3f439983657558cfb67629599',
+      'symbol': 'NKN',
+      'amount': '0.00000233',
+    },
+    {
+      'assetID': 'fc732edee1efdf968c23c20a9628eaa5a6ccb934',
+      'symbol': 'NNC',
+      'amount': '2000',
+    }
+  ]
+}
+
+Multiple account balances request sample
+<pre>
+// input
+{
+  params: [
+    {
+      address: 'AeysVbKWiLSuSDhg7DTzUdDyYYKfgjojru',
+    },
+    {
+      address: 'AbKNY45nRDy6B65YPVz1B6YXiTnzRqU2uQ',
+      assets: ['1578103c13e39df15d0d29826d957e85d770d8c9'],
+    },
+  ]
+}
+
+// output
+{
+  'AeysVbKWiLSuSDhg7DTzUdDyYYKfgjojru': [
+    {
+      'assetID': 'c56f33fc6ecfcd0c225c4ab356fee59390af8560be0e930faebe74a6daff7c9b',
+      'symbol': 'NEO',
+      'amount': '10',
+    },
+    {
+      'assetID': '602c79718b16e442de58778e148d0b1084e3b2dffd5de6b7b16cee7969282de7',
+      'symbol': 'GAS',
+      'amount': '777.0001',
+    },
+    {
+      'assetID': 'c36aee199dbba6c3f439983657558cfb67629599',
+      'symbol': 'NKN',
+      'amount': '0.00000233',
+    },
+    {
+      'assetID': 'fc732edee1efdf968c23c20a9628eaa5a6ccb934',
+      'symbol': 'NNC',
+      'amount': '2000',
+    }
+  ],
+  'AbKNY45nRDy6B65YPVz1B6YXiTnzRqU2uQ': [
+    {
+      'assetID': '1578103c13e39df15d0d29826d957e85d770d8c9',
+      'symbol': 'PHX',
+      'amount': '11000',
+    }
+  ]
+}
+</pre>
+
+'''Provider Request Handling'''
+
+Upon receiving this request, fetch the latest balance(s) for each given account, for the specific asset if applicable.
+Typechecking will be required on the input argument, as it can be eather a single BalanceRequest object, or an array of BalanceRequest objects. In the case where a specific asset is not provided to check the balance for, the wallet provider will fetch balances for all assets and tokens for that account.
+
+
+====getStorage====
+Reads the raw value in smart contract storage.
+
+
+'''Method Interface'''
+
+<pre>
+function getStorage({ scriptHash: string, key: string, network?: string }): Promise<StorageResponse>
+</pre>
+
+
+'''Input arguments'''
+
+<pre>
+interface GetStorageArgs {
+  scriptHash: string; // script hash of the smart contract to invoke a read on
+  key: string; // key of the storage value to retrieve from the contract
+  network?: string - Network to submit this request to. If omitted, will default to network the wallet is currently set to.
+}
+</pre>
+
+
+'''Success return value'''
+
+<pre>
+interface StorageResponse {
+  result: string; // The raw value that's stored in the contract
+}
+</pre>
+
+
+'''Error return value'''
+
+<pre>
+interface Error {
+  type: string; // `NO_PROVIDER`|`CONNECTION_REFUSED`|`RPC_ERROR`
+  description: string;
+  data: string;
+}
+</pre>
+
+
+'''Example'''
+
+<pre>
+getStorage({
+  scriptHash: '505663a29d83663a838eee091249abd167e928f5',
+  key: 'game.status'
```

Comment:
This is an example: looks like its passed as `ascii string`. But some languages may have difficulties on passing real string, for non-visible chars, like `\x00`, that may end-up closing strings by mistake on C++, at least.

Reply:
updated to specify hexstring format